DC-617 Update: Building your first custom VCC

eltesoro-koji opened the pull request:
New topic to replace "Building custom VCCs" topic.

This topic also requires two apps to be reviewed and listed in the app store:
-   Custom control scaffold: https://withkoji.com/@eltesorokoji/a5hv (Also scafcustctrl.koji-vccs.com/.)
-   Tester app: https://withkoji.com/@eltesorokoji/nb9e

(Marko already did a preliminary review of the code for both apps while debugging the "Loading ..." forever issue.)

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
++
+[source,bash]
+----
+npm install --save @withkoji/core
```
Does the custom control app require the core package? It doesn't seem like it should.

eltesoro-koji replied:
Removed. It was only needed by other things in the scaffold.

rasienko replied:
There are still lines in the code sample that rely on the core package (the import statement on line 2 and the Koji.ready method call on lines 44 - 49). If the Koji.ready method is really needed, we’ll need to put the installation info back. If not, we’ll remove the extra lines from the code sample and scaffold.

eltesoro-koji replied:
Restored the instructions to install @withkoji.core.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+Any app that uses your custom control must call `Koji.ui.capture.custom` and provide a `koji-vccs.com` subdomain or URL that points to the custom control.
+
+* You can determine the assigned URL of the custom control.
```
Suggest removing. If we recommend publishing on a subdomain of koji-vcc.com, let's focus on those instructions rather than providing every option in this tutorial.

eltesoro-koji replied:
Fixed.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+include::{includespath}/steps-domains.adoc[tag=click-register]
+====
+
+* You can also <<custom-domains#_using_a_domain_you_own,use a domain you own>>.
```
Same here. Suggest removing instructions/bullets for the other options and mentioning them with a link to the custom domains page.

eltesoro-koji replied:
Removed.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+* You can get your own subdomain of `koji-vccs.com`.
++
+.Learn how ...
+[%collapsible]
```
Since this will be the only instructions, you can remove it from the collapsible section. In general, limit collapsible sections as they are not searchable on the page and require additional clicks (things that developers tend not to like).

eltesoro-koji replied:
Fixed.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+include::{includespath}/steps-domains.adoc[tag=menu-app-custom-domain,click-add-new-domain,select-subdomain]
+
+include::{includespath}/steps-domains.adoc[tag=create-koji-vccs-com-subdomain]
```
Something wonky must be going on in the include file because several steps in this procedure are duplicated. Also, why are there so many includes? It's very hard to maintain, debug, and review when the content is constructed from many tiny pieces. Unless the reuse value is very high, consider writing the text directly in the page.

eltesoro-koji replied:
If something that could be reusable is not in an include file, the next person writing similar instructions in another topic would write the same step in their own way, which makes it harder to search and replace later on.

Even if Writer #2 creates an include at that point in time, someone will have to look for other instances of that step everywhere else to convert them into the new include. Again, an issue with search and replace. If they miss any, there would be multiple versions of the same step.

At a previous job, I have actually seen three instances of a piece of information that should be the same, but they all say different things. That feature changed many times and not all references to it were updated. The developer and the PM already left the company, and nobody knew which of those three is the correct information.

I'll look into compressing the steps. There are times when I want the full explanation and others when I just want the summary so I split up the step.

eltesoro-koji replied:
Found the cause of the duplicates: one of the end tags didn't match the start tag.

Also, there's now only one step for every screen that they see.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+==== Create a new project
+
+. Fork the https://withkoji.com/@eltesorokoji/a5hv/details/[Custom Control Scaffold].
```
This week's change to navigation makes the App Store is the primary way to use apps. So, we should update our tutorials to start from an app listing rather than the developer details or the app itself. When we decide on the starting point for this page, we can update this link.

eltesoro-koji replied:
Because the custom control scaffold isn't listed yet, I didn't have a way to point to an App Store listing.

I just got a message from Sean that he just fixed the issue. (It was a platform bug.) I'll test it, create a listing for both scaffold and tester app, then I'll update the links in the topic.

eltesoro-koji replied:
UPDATE: I hit another issue when I tried to create listings for both the custom control and the tester app. [Details in the Jira.](https://withkoji.atlassian.net/browse/DC-617?focusedCommentId=10385)

eltesoro-koji replied:
The scaffold and example tester app are now in the App Store and added to the ["For Developers" section](https://withkoji.com/create/for-developers) by Luis.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+==== Test your custom control
+
+. Create a standalone app.
```
Can we provide a working code sample for the consumer app? It is super-easy to add the custom control to the React Frontend Remix Scaffold: https://withkoji.com/apps/16fb355a-6471-44df-bf62-d75a4237c13e
We can provide instructions for starting from that app and adding the few lines of code needed to consume the custom control.
That way, someone could test/debug their own code. Since the linked test app is not listed in the App Store, the reader wouldn't be able to use it that way.

eltesoro-koji replied:
That's what the tester app is for. It's at the bottom of the section as an example. I'll move it up higher as an alternative to the React Scaffold.

eltesoro-koji replied:
UPDATE: Still trying to create the listing. [Details in the Jira.](https://withkoji.atlassian.net/browse/DC-617?focusedCommentId=10385)

eltesoro-koji replied:
The scaffold and example tester app are now in the App Store and added to the ["For Developers" section](https://withkoji.com/create/for-developers) by Luis.

rasienko commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+== Related resources
+
+* <<customvcc-reference#d>>
+* <<core-frontend-ui-capture#>>
```
Due to a bug in our custom parser, you must provide link text for this link. See DC-593 for more info.

eltesoro-koji replied:
* Added link texts for both.
* The "d" after the "#" was a typo. Fixed.

rasienko reviewed (changes requested):
Thanks for working through the bugs to get this draft ready. There are a couple of  general decisions we need to make about this article. 

First, do we want it to be a tutorial, like it was previously? Right now, it's less tutorial and more reference. I see benefits to both types of content, so it's really about what we want the goal of the page to be. Then, clarifying that goal in the intro so a reader knows what they'll be getting from this page.

Second, do we need to list anything in the store as a starting point, or can we provide instructions from an existing scaffold? I tend to prefer starting from an existing scaffold rather than adding more resources that we need to maintain.

eltesoro-koji commented:
> First, do we want it to be a tutorial, like it was previously? Right now, it's less tutorial and more reference. I see benefits to both types of content, so it's really about what we want the goal of the page to be. Then, clarifying that goal in the intro so a reader knows what they'll be getting from this page.
>
> Second, do we need to list anything in the store as a starting point, or can we provide instructions from an existing scaffold? I tend to prefer starting from an existing scaffold rather than adding more resources that we need to maintain.

1.
The scaffold takes the place of a tutorial. It's intended to be a working custom control. If we wrote a tutorial, it would essentially be, "Add this line here and that line there." So better to just add the lines for them, then they can focus on what makes their custom control different from other custom controls.

I'll add another paragraph in the intro.

2.
The scaffold and the tester app are major components of this topic. The scaffold needs to be added to the store; I'm just waiting for Sean to confirm that the "Loading..." issue is not in my code before I create a listing.

I forked this scaffold from the [React Frontend Remix Scaffold](https://withkoji.com/apps/16fb355a-6471-44df-bf62-d75a4237c13e), and I mentioned that in the README. I simply added the parts that are mentioned in the [Custom VCC Overview page](http://localhost:8000/reference/customvcc/withkoji-custom-vcc-sdk), so they don't have to do it themselves. It took me some time to figure out where each line should go, so having the scaffold would save approximately that much time multiplied by the number of developers who create custom controls.

If the process/information changes, we either edit the topic, the scaffold, or both. I think that would be approximately the same time spent. The advantage of the scaffold is time saved for the developer; not time savings for us.


FUTURE: 

* There's a way to pull "raw" code from GitHub to the docs, so that when the code is updated, the code block in the docs is updated too. But I'll have to ask Sean how that would work with our own Git repositories (or how to simulate something like that). 

* Eventually, I'll create custom control scaffolds forked from the basic scaffolds for [VueJS](https://withkoji.com/apps/0699f68c-37a3-45c1-8549-04f377ff8563) and [Svelte](https://withkoji.com/apps/e1a1c699-f227-4e30-aa06-72020a1b1258).

I'll create separate Jiras for these.

eltesoro-koji commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+Examples of custom controls are tile map editors, sound enhancers, or custom avatar creators, which enhance the app's interactivity.
+
+A custom control is an application in Koji; therefore, you create one the same way you create any other Koji application.
+However, unlike a typical Koji app, a custom control has only one context (Default).
```
Fixed.

eltesoro-koji commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+Because a custom control does not include other contexts, it cannot be used as a standalone app.
+
+== Creating a custom control
```
Fixed.

eltesoro-koji commented on src/asciidoc/docs/guides-general/build-custom-controls.adoc:
```diff
@@ -0,0 +1,210 @@
+
+----
+
+[.callout1]#Imports the `CustomVcc` class from the @withkoji/custom-vcc-sdk package.# +
```
Fixed.

rasienko commented:
> > First, do we want it to be a tutorial, like it was previously? Right now, it's less tutorial and more reference. I see benefits to both types of content, so it's really about what we want the goal of the page to be. Then, clarifying that goal in the intro so a reader knows what they'll be getting from this page.
> 
> The scaffold takes the place of a tutorial. It's intended to be a working custom control. If we wrote a tutorial, it would essentially be, "Add this line here and that line there." So better to just add the lines for them, then they can focus on what makes their custom control different from other custom controls.
> 
> I'll add another paragraph in the intro.
> 

Ok, makes sense to me! The intro looks good!

 > > Second, do we need to list anything in the store as a starting point, or can we provide instructions from an existing scaffold? I tend to prefer starting from an existing scaffold rather than adding more resources that we need to maintain.
> 
> The scaffold and the tester app are major components of this topic. The scaffold needs to be added to the store; I'm just waiting for Sean to confirm that the "Loading..." issue is not in my code before I create a listing. (UPDATE: Sean just fixed the issue.)
> 
> I forked this scaffold from the [React Frontend Remix Scaffold](https://withkoji.com/apps/16fb355a-6471-44df-bf62-d75a4237c13e), and I mentioned that in the README. I simply added the parts that are mentioned in the [Custom VCC Overview page](http://localhost:8000/reference/customvcc/withkoji-custom-vcc-sdk), so they don't have to do it themselves. It took me some time to figure out where each line should go, so having the scaffold would save approximately that much time multiplied by the number of developers who create custom controls.
> 
> If the process/information changes, we either edit the topic, the scaffold, or both. I think that would be approximately the same time spent. The advantage of the scaffold is time saved for the developer; not time savings for us.
> 

Ok, sounds good! Re: the problems with App Store listings, when I follow the links in the Jira, it looks like the “bundled app” hasn’t been identified in the listing. You would need to add the link to your published app in this field to link the listing with the app. If you’ve already tried that and it hasn’t saved the info, we might have a platform bug.

> FUTURE:
> 
> * There's a way to pull "raw" code from GitHub to the docs, so that when the code is updated, the code block in the docs is updated too. But I'll have to ask Sean how that would work with our own Git repositories (or how to simulate something like that).

It would be very cool to figure this out at some point!

> * Eventually, I'll create custom control scaffolds forked from the basic scaffolds for [VueJS](https://withkoji.com/apps/0699f68c-37a3-45c1-8549-04f377ff8563) and [Svelte](https://withkoji.com/apps/e1a1c699-f227-4e30-aa06-72020a1b1258).
> 
> I'll create separate Jiras for these.

These other scaffold are very low priorities and don't merit tickets. The React sample is sufficient to demonstrate the functionality in a straightforward and helpful way.

I believe the open items on this PR right now are:

- Whether we need the Core package in this tutorial/scaffold
- Getting the app listings into the store and updating the links in the page
- If you haven't already, getting a technical review

Otherwise, all of your changes look great to me. Let me know if I can help with any of the open items.

eltesoro-koji commented:
>     * Whether we need the Core package in this tutorial/scaffold

The scaffold definitely needs the core package. `Koji.ready` is required if listeners are called. It's not specific to custom controls. I thought that's what you meant when you said to take it out of the tutorial.

Your decision whether I put it back in the tutorial or not. But the scaffold would break without it.

eltesoro-koji commented:
> Re: the problems with App Store listings, when I follow the links in the Jira, it looks like the “bundled app” hasn’t been identified in the listing. You would need to add the link to your published app in this field to link the listing with the app.

Sean said the same in Slack on Saturday. The word "bundled" confused me because I thought it was an additional app, like the cable company *bundles* TV and internet together. I think how we use "bundled" and "template" in that page is confusing. I'll add "bundled" in our glossary discussion.

rasienko commented:
> The scaffold definitely needs the core package. `Koji.ready` is required if listeners are called. It's not specific to custom controls. I thought that's what you meant when you said to take it out of the tutorial.
> 
> Your decision whether I put it back in the tutorial or not. But the scaffold would break without it.

In that case, let's add it back to the tutorial. Sorry for the confusion!

rasienko commented:
> Sean said the same in Slack on Saturday. The word "bundled" confused me because I thought it was an additional app, like the cable company _bundles_ TV and internet together. I think how we use "bundled" and "template" in that page is confusing. I'll add "bundled" in our glossary discussion.

Great! Yes, it sounds like we could make that field name and the descriptions much clearer :)

eltesoro-koji commented:
> > The scaffold definitely needs the core package. `Koji.ready` is required if listeners are called. It's not specific to custom controls. I thought that's what you meant when you said to take it out of the tutorial.
> > Your decision whether I put it back in the tutorial or not. But the scaffold would break without it.
> 
> In that case, let's add it back to the tutorial. Sorry for the confusion!

Added back.

markof94 commented:
Looks pretty good to me, just a couple of things:

In the "Create a new project" section, there are steps which instruct how to install the @withkoji/core and @withkoji/custom-vcc-sdk packages, but since we're forking a project that already has those installed, maybe we can specify that it's already set up but those are the main packages that you need.
The steps themselves are correct but unnecessary in this case.

Sidenote - "--save" argument with the "npm install" command is not required and is on by default
https://blog.npmjs.org/post/161081169345/v500

eltesoro-koji commented:
> In the "Create a new project" section, there are steps which instruct how to install the @withkoji/core and @withkoji/custom-vcc-sdk packages, but since we're forking a project that already has those installed, maybe we can specify that it's already set up but those are the main packages that you need.
> The steps themselves are correct but unnecessary in this case.

Thanks much, Marko! 

I removed the steps and added a note in the next section that these packages are included in the scaffold and do not need to be reinstalled.

Making a note about "save" for future topics.

eltesoro-koji commented:
Per Slack message from Rachel:
>> Great! Quick notes on that change.... maybe add another sentence to the note indicating that both packages are required and a developer would need make sure they are installed if starting from another app. Also, maybe add links from the package names to the package overview pages.
Re: not needing --save, can you remove that from the package overview pages, also? Before we forget!

* Added the extra sentence about installing packages if using a different scaffold.
* Removed "--save" from the package overviews and a reusable.

rasienko commented:
Looks great! Merging now!
